staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -176,8 +178,9 @@ func newApfHandlerWithFilter(t *testing.T, flowControlFilter utilflowcontrol.Int
 		}))
 		apfHandler.ServeHTTP(w, r)
 		postExecute()
-		if atomicReadOnlyExecuting != 0 {
-			t.Errorf("Wanted %d requests executing, got %d", 0, atomicReadOnlyExecuting)
+		executing := atomic.LoadInt32(&atomicReadOnlyExecuting)
```

Comment:
The handler func here assumes that APF is done with the request by the time `apfHandler.ServeHTTP(w, r)` returns.  That was true until we added the special support for WATCH requests.   The real bug here is using this handler in tests of WATCH requests.

Reply:
Yeah - that was my point in the comment above - with this change it wouldn't be panicing but it would still be failing.

We need to fix the handler to ensure that that the request is finished by the time we exercise it.

Reply:
I think the `resultCh` already guarantee  that APF is done with the request by the time apfHandler.ServeHTTP(w, r) returns.
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go#L219

And `fakeWatchApfFilter` will ensure `execFn()` is called before panic. https://github.com/kubernetes/kubernetes/blob/fdb2d544751adc9fd2f6fa5075e9a16df7d352df/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go#L411-L419

Reply:
I think the `resultCh` already guarantee  that APF is done with the request by the time apfHandler.ServeHTTP(w, r) returns.
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go#L219

And `fakeWatchApfFilter` will ensure `execFn()` is called before panic. https://github.com/kubernetes/kubernetes/blob/fdb2d544751adc9fd2f6fa5075e9a16df7d352df/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go#L411-L419

Reply:
On second examination, I agree (but would cite different lines to show it).  In the post_execute_panic case, the send to resultCh at https://github.com/kubernetes/kubernetes/blob/fdb2d544751adc9fd2f6fa5075e9a16df7d352df/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go#L242 happens after the whole call/return of the `execute` defined at https://github.com/kubernetes/kubernetes/blob/fdb2d544751adc9fd2f6fa5075e9a16df7d352df/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go#L198-L216 .  The return from the APF handler happens after the receipt from resultCh, and the read of atomicReadOnlyExecuting happens after the return from the APF handler.  So I think I have just outlined the fact that the read happens after the write.